### PR TITLE
Simplify the network data source

### PR DESCRIPTION
**Pull request by dturner**
tl;dr the local data source is now the source of truth, the network data source has only 2 methods which load and save all data. 

One of the goals of this sample is to demonstrate how to deal with _different_ data sources but the `NetworkDataSource` interface was very similar to the local `TasksDao` interface. 

For this reason, I have simplified the `NetworkDataSource` interface to just 2 methods `loadTasks` and `saveTasks`, which load and save _all_ tasks respectively.

I've also changed the way that data synchronisation works inside `DefaultTasksRepository`: 

- When reading tasks they are obtained locally, unless `forceRefresh` is specified, in which case they are first obtained from the network data source
- When writing information they are written locally first, then written to the network data source

This is not a robust sync implementation. In a production app, if the remote write failed and a force refresh was performed the local data would be overwritten. But I think it's better than what we had before which was essentially random behaviour.

### CODE REVIEW

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt [hunk 1]
```diff
@@ -95,68 +83,51 @@ class DefaultTasksRepository(
         }
     }
 
-    override fun getTaskStream(taskId: String): Flow<Task?> {
-        return tasksDao.observeTaskById(taskId).map { it.toExternalModel() }
+    private suspend fun sendTasksToNetworkDataSource() {
```
Comment (manuelvicnt) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt:86]: `sendTasksToNetworkDataSource` sounds too prescriptive with what the network is doing. Maybe we should choose a more generic name? We shouldn't know the network data source impl details.

Maybe something something shorter and more generic like `syncNetworkTasks` would do

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt [hunk 2]
```diff
@@ -95,68 +83,51 @@ class DefaultTasksRepository(
...
+        // Real apps may want to use a proper sync strategy here to avoid data conflicts.
+        val localTasks = tasksDao.getTasks()
+        tasksNetworkDataSource.saveTasks(localTasks.toNetworkModels())
```
Comment (manuelvicnt) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt:89]: Same comment applies to all places where we do mapping.

We don't know how expensive retrieving from the local data source and mapping the models is. We should move the execution to the `defaultDispatcher`. We could be talking about hundreds / thousands of tasks and that could block the calling thread.

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt (hunk 1 above)
Comment (dturner) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt:86]: I agree this name is too verbose, however, the direction of data travel and where it's going to is important. For this reason `sync` is too generic. How about just `sendTasksToNetwork`?

Comment (dturner) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt:86]: btw I chose this name to match the existing counterpart method `updateTasksFromNetworkDataSource`, I could also rename this to `loadTasksFromNetwork`.

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt (hunk 2 above)
Comment (dturner) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt:89]: Good point. I've moved execution to `Dispatchers.Default` for all *bulk* network, local db and mapping operations. Do you think it's worth doing this for single operations as well, like `getTask`? To me that seems like overkill.

Comment (dturner) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/DefaultTasksRepository.kt]: Just to note that only bulk network and mapping operations need `withContext`, Room will take care of using the correct dispatcher for local database operations.

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt [hunk 3]
```diff
@@ -18,141 +18,128 @@ package com.example.android.architecture.blueprints.todoapp.data
...
     }
 
     override suspend fun refreshTask(taskId: String) {
-        updateTaskFromRemoteDataSource(taskId)
+        loadTasksFromNetwork()
     }
 
-    private suspend fun updateTasksFromRemoteDataSource() {
-        val remoteTasks = tasksNetworkDataSource.loadTasks()
+    private suspend fun loadTasksFromNetwork() {
```
Comment (manuelvicnt) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt:84]: Nit: move private functions below public overriden ones

**Review by manuelvicnt — APPROVED** [review]

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt (hunk 3 above)
Comment (dturner) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt]: Done

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt [hunk 4]
```diff
@@ -18,141 +18,128 @@ package com.example.android.architecture.blueprints.todoapp.data
...
     override suspend fun refreshTasks() {
-        updateTasksFromRemoteDataSource()
+        loadTasksFromNetwork()
     }
 
     override fun getTasksStream(): Flow<List<Task>> {
         return tasksDao.observeTasks().map { tasks ->
-            tasks.map { task ->
-                task.toExternalModel()
+            withContext(coroutineDispatcher) {
```
Comment (tunjid) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt:74]: Why do we need to change the context? `tasks.toExternalModels()` seems synchronous

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt [hunk 5]
```diff
@@ -18,141 +18,128 @@ package com.example.android.architecture.blueprints.todoapp.data
...
+            tasksDao.deleteTasks()
+            remoteTasks.forEach { task ->
+                tasksDao.insertTask(task.toTaskEntity())
+            }
+        }
+    }
+
+    private suspend fun saveTasksToNetwork() {
+        withContext(coroutineDispatcher) {
+            // Real apps may want to use a proper sync strategy here to avoid data conflicts.
```
Comment (tunjid) [inline app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt:140]: May also want to add that this may be work best deferred to `WorkManager`.

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt (hunk 4 above)
Comment (dturner) [reply app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/DefaultTasksRepository.kt:74]: For exactly that reason, it is synchronous and could therefore block the calling thread indefinitely if it was, say, mapping a million tasks
